### PR TITLE
Improving the performance and generality of adaptation

### PR DESCRIPTION
**Pull request by mohamed82008**
This PR:
1. Enables full-fledged arbitrary precision handling, e.g. Float64 or Float32, 
2. Resolves some type instability issues from using abstract type fields, and
3. Reduces allocations in a few places.

### CODE REVIEW

**Comment by mohamed82008** [comment]
@xukai92 there is no functional change in this PR AFAIK, but the tests are failing. Any idea why?

**Comment by mohamed82008** [comment]
It seems that tests pass now!

File: src/adaptation/precond.jl [hunk 1]
```diff
@@ -29,38 +30,42 @@ function reset!(nc::NaiveVar{T}) where {T<:AbstractFloat}
...
 function add_sample!(wv::WelfordVar, s::AbstractVector)
     wv.n += 1
-    δ = s .- wv.μ
-    wv.μ .+= δ ./ wv.n
-    wv.M .+= δ .* (s .- wv.μ)
+    @unpack μ, M, n = wv
+    for i in eachindex(s)
+        δ = s[i] - μ[i]
+        μ[i] += δ / n
+        M[i] += δ * (s[i] - μ[i])
```
Comment (yebai) [inline src/adaptation/precond.jl]: Why switching from broadcasting to loop? Is this for readability or for efficiency?

Comment (mohamed82008) [reply src/adaptation/precond.jl]: Efficiency to avoid allocating `δ`.

**Comment by mohamed82008** [comment]
`Float64` is the default even on 32-bit Julia, that is `typeof(1.0) === Float64` on both 32-bit and 64-bit Julia.

**Comment by yebai** [comment]
Thanks!

**Pull request merged**